### DISCUSSION TITLE
[Proposal]: Static Extension methods for static classes

### DISCUSSION THREAD

**Discussion by MadL1me** [dotnet/csharplang#5749] [Language Ideas] [answered]
# Static extension methods for static/non-static classes

- [x] Proposed
- [ ] Prototype: Not Started
- [ ] Implementation: Not Started
- [ ] Specification: Not Started

# Summary
Allow to make static extension methods for static/non-static classes.

# Motivation
Is some use cases, when you have a library or API with static classes, if you use that static class you want to use it anywhere without adding new wrapper-classes and other facade-like abstractions when static class doesn't have methods/api you want. But you do not have an ability to do so:

```cs

var isDecember = Calendar.IsDecember(DateTime.Now);
        
// I want to use somehow similar method which not represented in Calendar static class
var isNomber = Calendar.IsNovember(DateTime.Now);  // error, method does not exists

public static class Calendar
{ 
    public static bool IsDecember(DateTime dateTime)
    {
        return dateTime.Month == 12;
    }
}
```

This is not an issue, because you can extend static classes by making them partial: 

```cs

var isDecember = Calendar.IsDecember(DateTime.Now);
var isNomber = Calendar.IsNovember(DateTime.Now);  // yay, now it works!


public static partial class Calendar
{ 
    public static bool IsDecember(DateTime dateTime)
    {
        return dateTime.Month == 12;
    }
}

public static partial class Calendar
{ 
    public static bool IsNovember(DateTime dateTime)
    {
        return dateTime.Month == 11;
    }
}
```

**But the problem is - not a lot of people use ```partial``` keyword, and 99% of the time you have static class without it being partial.
Moreover - most static classes in BCL are just static and not partial - ```System.Math``` as an example.**

# Detailed design
We can introduce ```static``` keyword used in a method declaration: 

```cs 
public static class CalendarExtensions
{
    public static bool IsNovember(this static Calendar, DateTime dateTime) 
    {
        return dateTime.Month == 11;
    }
}
```

Now we can extend methods in static classes!

```cs

var isDecember = Calendar.IsDecember(DateTime.Now);
var isNomber = Calendar.IsNovember(DateTime.Now); 

public static class Calendar
{ 
    public static bool IsDecember(DateTime dateTime)
    {
        return dateTime.Month == 12;
    }
}

public static class CalendarExtensions
{ 
    public static bool IsNovember(this static Calendar, DateTime dateTime)
    {
        return dateTime.Month == 11;
    }
}
```

And also now we basically can make "static" extension methods even for non static classes (we'll have 2 different ways to make static classes: 

```cs
class A 
{
    public void B() 
    {
        // do stuff
    }
}

class AExtensions
{
    public static void C(this A a) // default extension method
    {
        // do stuff
    }

    public static void D(this static A) // new static extension method
    {
        // do other stuff
    }
}

var a = new A();
a.B() ;
a.C(); // default extension method

A.D(); // static extension mehod! Now we don't even need to allocate new instance 

```

(P.S - It's my first try to give constructive issue and proposal for big project, so please, if I did something wrong - tell me so I can improve in my future issues. Thanks!)

**Comment by fa1ntl1ght** [comment, marked as answer]
#5497 includes this (with a different syntax)

**Comment by CyrusNajmabadi** [comment]
> (P.S - It's my first try to give constructive issue and proposal for big project, so please, if I did something wrong - tell me so I can improve in my future issues. Thanks!)

Welcome! You've done nothing wrong!

About all we might say is that it's good to look if there are potential duplicate discussions on these topics already.  If you're not sure, you can also come to our discord to run the idea past peoplt there to see if they know of a preexisting topic.
